### ISSUE TITLE
New rule proposal: no-unreachable-loop

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12381]
**Please describe what the rule should do:**

Disallows loops which body exits the loop in all paths and therefore can never even reach the test condition for the second time (or first time if it's `do-while`), meaning that it isn't actually a loop.

Targets `while`, `do-while`, `for`, `for-in` and `for-of` loops.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
while (foo) {
    doSomething(foo);
    foo = foo.parent;
    break;
}

for (i = 0; i < arr.length; i++) {
    if (isSomething(arr[i])) {
        break;
    } else {
        return false;
    }
}

do {
    if (foo.type === "bar") {
        return foo;
    } else {
        throw new Error("invalid");
    }
} while (foo);

for (foo of bar) {
    if (foo.type === "bar") {
        doSomething(foo);
    }
    break;
}

for (foo in bar) break;
```

The code is correct only if the end of the body is reachable (i.e. loops from the end) or the loop has a `continue` statement.

**Why should this rule be included in ESLint (instead of a plugin)?**

It's a very possible error in the code, something is missing or something else is at the wrong place.

Even if it isn't an error the code should be refactored to avoid unnecessary loop syntax.

This is a similar type of error as `no-unreachable`, but I think this should be a separate rule because there is no particular unreachable code in these cases (except for the code in `do-while` test and `for-loop` update, but I think that `no-unreachable` doesn't report that).

In some cases both `no-unreachable` and this rule can report warnings (e.g. `no-unreachable` for code after the last `break`, this rule for the whole loop), but I think that isn't overlapping.

**Are you willing to submit a pull request to implement this rule?**

Yes, I'd love to work on this.

**Comment by ljharb** [comment]
For `while` or `do` or `for`, this makes sense to me - but `for..in` is occasionally (much more often historically) used like this:
```js
for (var k in obj) {
  k = true;
  break;
}
if (k) {
  // object has enumerable properties
}
```

The same could be true of `for..of` - both `for..in` and `for..of` are *not* inherently unnecessary here since they perform iterations.

**Comment by mdjermanovic** [comment]
That's correct, I think there should be options to turn off `for-in` and `for-of` checks for this reason.

An alternative for an user is to extract that code to a function (`hasEnumerableProperties` or `getFirstEnumerableName`) and eslint-disable just that one line, or instead use `Object.keys` over prototype chain in the function. Similar for `for-of`, an alternative could be to access the iterator directly, or do `[first] = foo` if `undefined` isn't an expected value. But, all of this would be probably inconvenient for an existing codebase.

I'd still vote that the rule checks `for-in` and `for-of` as it could catch some real errors, but behind an option (or options).

**Comment by ljharb** [comment]
Most configurable would be an option for each construct, regardless of the defaults.

**Comment by mdjermanovic** [comment]
I agree, one option for `for-in` and another for `for-of`.

Or did you mean all 5 options, `for`, `while` and `do` as well?

**Comment by ljharb** [comment]
Each option separately.

**Comment by mdjermanovic** [comment]
Any thoughts from the team about this rule? :-)

It looks to me that using the already existing code path analysis features this rule could reliably (without false negatives or false positives) detect these bugs in the code, and might be a nice complement to the `no-unreachable` rule.

**Comment by kaicataldo** [comment]
@ilyavolodin Mind clarifying if your thumb on the comment above is in support of this proposal? If so, we can mark this as accepted :)

**Comment by mdjermanovic** [comment]
One more :+1: to accept, if thumb on the comment doesn't count :)

**Comment by mdjermanovic** [comment]
I'm working on this.
